tianshou/policy/modelfree/fqf.py

```diff
@@ -73,15 +74,16 @@ def _target_q(self, buffer: ReplayBuffer, indices: np.ndarray) -> torch.Tensor:
         next_dist = next_dist[np.arange(len(act)), act, :]
         return next_dist  # shape: [bsz, num_quantiles]
 
-    def forward(
+    # TODO: add protocol type for return, fix Liskov substitution principle violation
+    def forward(  # type: ignore
         self,
-        batch: Batch,
+        batch: RolloutBatchProtocol,
         state: Optional[Union[dict, Batch, np.ndarray]] = None,
         model: str = "model",
         input: str = "obs",
         fractions: Optional[Batch] = None,
         **kwargs: Any,
-    ) -> Batch:
+    ) -> BatchProtocol:
```

Comment:
Made a new protocol. 

```python
class FQFBatchProtocol(ModelOutputBatchProtocol):
    fractions: torch.Tensor
    quantiles_tau: torch.Tensor
```

In one of the last commits I renamed the `ActionBatchProtocol` to `ModelOutputBatchProtocol`, since we need the `ActBatchProtocol` with only the actions field (used e.g. in imitation learning)